src/transformStateWithClones.js

```diff
@@ -7,7 +7,63 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const arrayResults = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        if (arrayResults.length === 0) {
+          arrayResults.push(Object.assign({}, state, action.extraData));
+        } else {
+          arrayResults.push(
+            Object.assign(
+              {},
+              arrayResults[arrayResults.length - 1],
+              action.extraData,
+            ),
+          );
+        }
```

Comment:
The logic for checking `arrayResults` length and creating a new state object is repeated. Consider refactoring to eliminate this duplication and improve maintainability.